src/enumerate.js

```diff
@@ -0,0 +1,188 @@
+        }]
+    };
+
+    if (_params.filter) {
+        res['s:Body']['wsen:Enumerate'][0]['wsman:Filter'] = [{
+            '@': {
+              'Dialect': _params.filterDialect || 'http://www.w3.org/TR/1999/REC-xpath-19991116'
+            },
+            '#':  _params.filter
+        }];
+    }
+
+    return js2xmlparser.parse('s:Envelope', res);
+}
+
+function constructPullEnumerationRequest(_params) {
+    var res = winrm_soap_req.getSoapHeaderRequest({
+        'resource_uri': _params.resourceUri || 'http://schemas.dmtf.org/wbem/wscim/1/*',
+        'action': 'http://schemas.xmlsoap.org/ws/2004/09/enumeration/Pull'
+    });
+
+    res['s:Body'] = {
+        'wsen:Pull': [{
+            'wsen:EnumerationContext': _params.enumerationId,
+            'wsen:MaxTime': _params.maxTime || 'PT10S',
+            'wsen:MaxElements': _params.maxElements || 20
+        }]
+    };
+
+    return js2xmlparser.parse('s:Envelope', res);
+}
+
+function constructReleaseEnumerationRequest(_params) {
+    var res = winrm_soap_req.getSoapHeaderRequest({
+        'resource_uri': _params.resourceUri || 'http://schemas.dmtf.org/wbem/wscim/1/*',
+        'action': 'http://schemas.xmlsoap.org/ws/2004/09/enumeration/Release'
+    });
+
+    res['s:Body'] = {
+        'wsen:Release': [{
+            'wsen:EnumerationContext': _params.enumerationId
+        }]
+    };
+
+    return js2xmlparser.parse('s:Envelope', res);
+}
+
+function getObjects(items) {
+    let itemObjects = [];
+    if (items) {
+        for (let item of items[0][Object.keys(items[0])[0]]) {
+            let itemObject = {};
+            // TODO return map of multiple item keys?
+            Object.keys(item).forEach((k) => {
+                if (k === '$') { return; }
+                let keyName = k.replace(/^p:/, '');
+                let value = item[k][0];
+                if (value && value['Datetime']) {
+                    value = value['Datetime'][0];
+                }
+                if (value && value['$']) {
+                    if (value['$']['xsi:nil'] === 'true') {
+                        value = null;
+                    }
+                }
+                itemObject[keyName] = value;
+            });
+            itemObjects.push(itemObject);
+        }
+    }
+    return itemObjects;
+}
+
+module.exports.doBeginEnumeration = async function (_params) {
+    var req = constructBeginEnumerationRequest(_params);
+
+    var auth = _params.auth;
+    if (_params.authOnce) {
+        auth = typeof _params.authOnce === 'string' ? _params.authOnce : _params.auth;
+        _params.auth = undefined;
+        _params.authOnce = undefined;
+    }
+    var result = await winrm_http_req.sendHttp(req, _params.host, _params.port, _params.path, auth, _params.agent);
+
+    if (result['s:Envelope']['s:Body'][0]['s:Fault']) {
+        var faultCode = result['s:Envelope']['s:Body'][0]['s:Fault'][0]['s:Code'][0]['s:Subcode'][0]['s:Value'][0];
```

Comment:
are these structures always guaranteed to be correct? looks like these could be prone to a 'can't read property ? of undefined` error...

Reply:
I think `s:Subcode` should always exist for this, but yep, the spec says s:Reason is optional. I'll check them both and return something regardless.